File: 백준/Silver/1874. 스택 수열/스택 수열.py
sequence = list()
n = int(input())

for _ in range(n):
    sequence.append(int(input()))
    
def stack_sequence(n, sequence):
    # 스택에 push하는 순서는 반드시 오름차순
    stack_seq = []
    # 연산자를 담기 위한 배열
    operators = []
    # 현재 처리된 수열의 인덱스
    sequence_index = 0
    
    for i in range(1, n + 1):
        stack_seq.append(i)
        operators.append('+')
        # 스택의 peek 부분과 현재 처리할 수열의 값을 비교
        while stack_seq and stack_seq[-1] == sequence[sequence_index]:
            stack_seq.pop()
            sequence_index += 1
            operators.append('-') # 두 수가 일치할 경우 pop
            
    if stack_seq:
        print("NO")
    else:
        for o in operators:
            print(o)

stack_sequence(n, sequence)            
    

File: 백준/Silver/1874. 스택 수열/스택 수열.js
let fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

let n = Number(input[0]);
let sequence = [];

for (let i = 1; i < input.length; i++) {
    if (input[i] !== '') {        
        sequence.push(Number(input[i]))
    }
}

function stack_sequence(n, sequence) {
    
    let stack_seq = []
    let operators = []
    let sequence_index = 0
    
    for (let i = 1; i <= n; i++) {
        stack_seq.push(i);
        operators.push('+');     
        while(
          stack_seq.length &&
            stack_seq[stack_seq.length - 1] === sequence[sequence_index]
        ) {
          stack_seq.pop();
          sequence_index++;
          operators.push('-')
        }       
    }
    
    if (stack_seq.length) {
        console.log("NO");
    } else {
        console.log(operators.join("\n"))
    }
}

stack_sequence(n, sequence);




